zerver/lib/actions.py

```diff
@@ -815,7 +816,17 @@ def do_send_messages(messages_maybe_none):
                     um.flags |= UserMessage.flags.is_me_message
 
                 user_message_flags[message['message'].id][um.user_profile_id] = um.flags_list()
-            ums.extend(ums_to_create)
+
+            user_messages = []
+            for um in ums_to_create:
+                if (um.user_profile.long_term_idle and
+                        um.message.recipient.type == Recipient.STREAM and
+                        not um.flags.mentioned.is_set and
+                        not um.flags.is_me_message.is_set):
+                    continue
```

Comment:
I think it might be more robust to make the `mentioned` checks be just `um.flags == 0`?